### DISCUSSION TITLE
Consider making ConsumerDispatchConcurrency a per-consumer setting instead of a per-connection setting

### DISCUSSION THREAD

**Discussion by kharabasz** [rabbitmq/rabbitmq-dotnet-client#1349] [General]
### Is your feature request related to a problem? Please describe.

Currently if we want concurrent processing we need to set ConsumerDispatchConcurrency on the connection to > 1 (preferably in an AsyncBasicConsumer). This value can be tuned together with the Prefetch count to achieve desired throughput. 

If your program contains multiple consumers and you want to have each configured with a different throughout then ConsumerDispatchConcurrency should (in theory) be set to the max of what is needed on each individual consumer as it is a global on the connection. There can also be combinations that are not possible without using a separate connection per consumer.

### Describe the solution you'd like

Have ConsumerDispatchConcurrency be a setting on the Consumer instead of on the Connection.

### Additional context

Please let me know if I am missing some knowledge related to usage, such that it is preferred to use multiple connections per program to isolate consumers.

**Comment by michaelklishin** [comment]
I am highly skeptical that this supposedly fine-grained control will actually work as expected.

The underlying concurrency are not per-consumer and neither is prefetch by default and in most systems.

Concurrency controls are meant to
limit the maximum concurrency rate, not control consumer rate or concurrency rate. You can achieve that in your own concurrency implementation by dispatching to any thread pool or executor you need.

In other clients doing so would mean maintaining a thread pool or executor per consumer, which would be a complete overkill.

If some of your consumers are high throughput and others are not, use two consumer connections, or three, or five. If you need more than that, your concurrency controls will likely only provide intellectual satisfaction and not any resource efficiency gains.

**Reply by kharabasz** [reply]
Before I ask some follow-up questions - should I be looking at main for the latest version of the code? I'm currently assuming main will eventually ship as 7.x.

**Reply by lukebakken** [reply]
> should I be looking at main for the latest version of the code? I'm currently assuming main will eventually ship as 7.x.

Yep, that's correct.

**Reply by kharabasz** [reply]
Would you help me to understand how this latest code from main works? [ConsumerDispatcherChannelBase.cs#L19-L45](https://github.com/rabbitmq/rabbitmq-dotnet-client/blob/main/projects/RabbitMQ.Client/client/impl/ConsumerDispatching/ConsumerDispatcherChannelBase.cs#L19-L45)
 ```
        protected ConsumerDispatcherChannelBase(ChannelBase channel, int concurrency)
        {
            _channel = channel;
            var workChannel = Channel.CreateUnbounded<WorkStruct>(new UnboundedChannelOptions
            {
                SingleReader = concurrency == 1,
                SingleWriter = false,
                AllowSynchronousContinuations = false
            });
            _reader = workChannel.Reader;
            _writer = workChannel.Writer;
            
            Func<Task> loopStart = ProcessChannelAsync;
            if (concurrency == 1)
            {
                _worker = Task.Run(loopStart);
            }
            else
            {
                var tasks = new Task[concurrency];
                for (int i = 0; i < concurrency; i++)
                {
                    tasks[i] = Task.Run(loopStart);
                }
                _worker = Task.WhenAll(tasks);
            }
        }
```
As far as I understand for the async case, a concrete `AsyncConsumerDispatcher` is created for every channel and its' concurrency is initialized to the connection's `ConsumerDispatchConcurrency` setting. If the concurrency is > 1, then the `AsyncConsumerDispatcher` will spin up `N = concurrency` readers on the workChannel. 

I'm assuming that this works in such a way that for a single consumer with prefetch >= concurrency that as long as the queue has messages those readers will together roughly process N messages in parallel as allowed by the task scheduler.

I'm assuming that setting prefetch > concurrency is only used in cases where message sizes are large and we want to pre-pull messages into the workChannel to save on network transfer times?

I'm assuming that starting 2 consumers: one with concurrency = prefetch and one with concurrency = prefetch/2, would lead to the second consumer running prefetch/2 readers that would never see a message?

> I am highly skeptical that this supposedly fine-grained control will actually work as expected.
> 
> The underlying concurrency are not per-consumer and neither is prefetch by default and in most systems.
>
> Concurrency controls are meant to limit the maximum concurrency rate, not control consumer rate or concurrency rate. You can achieve that in your own concurrency implementation by dispatching to any thread pool or executor you need.

I should not have used `throughput` in the original discussion request. What I did mean is limiting concurrency rate per consumer.

> In other clients doing so would mean maintaining a thread pool or executor per consumer, which would be a complete overkill.
> 
> If some of your consumers are high throughput and others are not, use two consumer connections, or three, or five. If you need more than that, your concurrency controls will likely only provide intellectual satisfaction and not any resource efficiency gains.

Do you feel the same way if we are on the same page and discussing limiting concurrency rate?

Web services in our current set-up can serve requests via http or amqp. They have defined apis for both - we like grouping consumer logic into a single service by domain which gives us services that have multiple consumers running, at times needing a different concurrency limit.

We have 2 versions of a thin wrapper around the rabbitmq-dotnet-client:
- In the latest version we abstract a single connection away behind an `ObjectPool<IModel>` and wire up a single publisher channel and multiple consumer channels per service. The ObjectPool also exposes a HealthCheck on the connection.
- In the previous version we wire up a single publisher channel and multiple consumer channels per service, with each having it's own connection.

In the context of the first approach, would moving `ConsumerDispatchConcurrency` to the consumer level or adding a `ConsumerDispatchConcurrency` override at the consumer level make sense? 
Is there a different way to achieve the same result with a single connection?
Should we just revert to the many connection approach from our previous wrapper?

**Reply by lukebakken** [reply]
> Should we just revert to the many connection approach from our previous wrapper?

Yes, mainly because I don't think we will change the dispatch concurrency behavior in version 7. Maybe in a future version if multiple requests for the feature are made.

Please take the time to test the alpha releases for version 7, it would be a _MAJOR_ contribution to do so:

https://www.nuget.org/packages/RabbitMQ.Client/7.0.0-alpha.6

Thank you @kharabasz
